clients/admin-ui/src/features/common/system-data-flow/DataFlowAccordionForm.tsx

```diff
@@ -0,0 +1,223 @@
+  useToast,
+} from "@fidesui/react";
+import { FormGuard } from "common/hooks/useIsAnyFormDirty";
+import { GearLightIcon } from "common/Icon";
+// import QuestionTooltip from "common/QuestionTooltip";
+import { DataFlowSystemsDeleteTable } from "common/system-data-flow/DataFlowSystemsDeleteTable";
+import DataFlowSystemsModal from "common/system-data-flow/DataFlowSystemsModal";
+import { errorToastParams, successToastParams } from "common/toast";
+import { Form, Formik, FormikHelpers } from "formik";
+import React, { useEffect, useMemo, useState } from "react";
+
+import { useAppSelector } from "~/app/hooks";
+import {
+  useGetAllSystemsQuery,
+  useUpdateSystemMutation,
+} from "~/features/system";
+import { selectAllSystems } from "~/features/system/system.slice";
+import { DataFlow, System } from "~/types/api";
+
+const defaultInitialValues = {
+  dataFlowSystems: [] as DataFlow[],
+};
+
+export type FormValues = typeof defaultInitialValues;
+
+type DataFlowAccordionItemProps = {
+  isIngress?: boolean;
+  system: System;
+  isSystemTab?: boolean;
+};
+
+export const DataFlowAccordionForm = ({
+  system,
+  isIngress,
+  isSystemTab,
+}: DataFlowAccordionItemProps) => {
+  const toast = useToast();
+  const flowType = isIngress ? "Source" : "Destination";
+  const pluralFlowType = `${flowType}s`;
+  const dataFlowSystemsModal = useDisclosure();
+  const [updateSystemMutationTrigger, mutationResult] =
+    useUpdateSystemMutation();
+
+  useEffect(() => {
+    if (mutationResult.isLoading || mutationResult.isUninitialized) {
+      return;
+    }
+
+    if (mutationResult.isError) {
+      toast(errorToastParams("Failed to update data flows"));
+    }
+    if (mutationResult.isSuccess) {
+      toast(successToastParams(`${pluralFlowType} updated`));
+    }
+  }, [
+    mutationResult.error,
+    mutationResult.isError,
+    mutationResult.isLoading,
+    mutationResult.isSuccess,
+    mutationResult.isUninitialized,
+    pluralFlowType,
+    toast,
+  ]);
+
+  useGetAllSystemsQuery();
+  const systems = useAppSelector(selectAllSystems);
+
+  const initialDataFlows = useMemo(() => {
+    let dataFlows = isIngress ? system.ingress : system.egress;
+    if (!dataFlows) {
+      dataFlows = [];
+    }
+    const systemFidesKeys = systems ? systems.map((s) => s.fides_key) : [];
+
+    return dataFlows.filter((df) => systemFidesKeys.includes(df.fides_key));
+  }, [isIngress, system, systems]);
+
+  const [assignedDataFlow, setAssignedDataFlows] =
+    useState<DataFlow[]>(initialDataFlows);
+
+  useEffect(() => {
+    setAssignedDataFlows(initialDataFlows);
+  }, [initialDataFlows]);
+
+  const handleSubmit = async (
+    { dataFlowSystems }: FormValues,
+    { resetForm }: FormikHelpers<FormValues>
+  ) => {
+    const updatedSystem = {
+      ...system,
+      ingress: isIngress ? dataFlowSystems : system.ingress,
+      egress: !isIngress ? dataFlowSystems : system.egress,
+    };
+    await updateSystemMutationTrigger(updatedSystem);
```

Comment:
I didn't totally understand your comment about the older closure—why can't we grab the result from the await here and then handle any errors right afterwards? 

like this pattern: https://github.com/ethyca/fides/blob/78448d9ff4d04fd279087e3889c4c728a86b1e89/clients/admin-ui/src/features/config-wizard/AuthenticateAwsForm.tsx#L102-L115

Reply:
ahh,  thanks for sharing this. It worked great.

I checked what was returned by the mutation and saw it only return an object like this 

```json
{ "data": "..."}
```

so I assumed that it only returned that. I thought it would the normal result object that as `isError` and `isSuccess`.


The actually return type is something like `{"data": any} | {"error": any}`. I didn't realize that.